[wip/rfc][dynamo] Flag to skip all guards

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134174
* #134181
* #134039
* #134016
* #133742

This is a requested feature by the users. What i have seen is users want something like `export` but not really `export`. They want the user friendliness of torch.compile but tight control over the fixed overheads.

The users are aware of the risks here and the responsibility/danger that comes with this flag. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

anijain2305 committed:
[dynamo] Flag to skip all guards

[ghstack-poisoned]

youkaichao commented:
Thanks for the efforts! cc @woosukkwon we can test this with vLLM + TPU, since we already use torch nightly.

youkaichao commented:
To be more precise, what we want would be customizing guards. If we disable all guards (as in this PR), then later compilation will directly hit the guard, and run the compiled code directly, right?

Ideally, we have several compilations, and we'd like to control the guards we need to check.

For example, we'd like to use `torch.compile` to compile three times:

profile phase:
execute model with fake input, to estimate the space of memory available for kvcache

prefill phase: 
input shape [1, prefill_len, model_size]
prefill_len padded to the nearest power of 2
execute prefill for each request separately.

decode phase:
input shape [batchsize, 1, model_size]
batchsize padded to multiples of 8

After compiling the graph for prefill, we'd like to customize the guards to only contain shape check for the input.

anijain2305 committed:
Update on "[dynamo] Flag to skip all guards"


This is a requested feature by the users. What i have seen is users want something like `export` but not really `export`. They want the user friendliness of torch.compile but tight control over the fixed overheads.

The users are aware of the risks here and the responsibility/danger that comes with this flag. 


[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Flag to skip all guards"


This is a requested feature by the users. What i have seen is users want something like `export` but not really `export`. They want the user friendliness of torch.compile but tight control over the fixed overheads.

The users are aware of the risks here and the responsibility/danger that comes with this flag. 


[ghstack-poisoned]

youkaichao commented on torch/_dynamo/guards.py:
```diff
@@ -2877,10 +2877,19 @@ def install_guard(*guards, skip=0):
+        add = TracingContext.get().guards_context.dynamo_guards.add
+        for guard in guards:
+            assert isinstance(guard, Guard)
+            add(guard, collect_debug_stack=collect_debug_stack, skip=skip + 1)
```
I tried locally to comment out this line, but I still get some guards:

```python
[rank0]:V0822 00:17:15.035000 626512 torch/_dynamo/guards.py:2704] [0/1] [__recompiles_verbose] Recompiling function forward in /home/youkaichao/vllm/vllm/worker/tpu_model_runner.py:599
[rank0]:V0822 00:17:15.035000 626512 torch/_dynamo/guards.py:2704] [0/1] [__recompiles_verbose]     triggered by the following guard failure(s):
[rank0]:V0822 00:17:15.035000 626512 torch/_dynamo/guards.py:2704] [0/1] [__recompiles_verbose]     guard 0 failures:
[rank0]:V0822 00:17:15.035000 626512 torch/_dynamo/guards.py:2704] [0/1] [__recompiles_verbose]     - 0/0: Eq(Mod(L['token_ids'].size()[1], 16), 0)                      # _dynamo/output_graph.py:455 in init_ambient_guards; 2 <= L['token_ids'].size()[1]                                 # _dynamo/output_graph.py:455 in init_ambient_guards; 2 <= L['attn_metadata'].slot_mapping.size()[1]                # _dynamo/output_graph.py:455 in init_ambient_guards
```

is it related to some 0/1 specialization? I vaguely remember some related terms from @ezyang 's podcast.

anijain2305 replied:
Ah. this is not the only place we are inserting guards. I will have to sprinkle this flag at more places then.

anijain2305 replied:
Try the latest commit.

youkaichao replied:
then I get error:

```python
[rank0]:   File "<eval_with_key>.1 from /home/youkaichao/miniconda3/envs/myenv/lib/python3.10/site-packages/torch/fx/experimental/proxy_tensor.py:910 in wrapped", line 9, in forward
[rank0]:   File "/home/youkaichao/miniconda3/envs/myenv/lib/python3.10/site-packages/torch/_ops.py", line 720, in __call__
[rank0]:     return self._op(*args, **kwargs)
[rank0]: RuntimeError: shape '[1]' is invalid for input of size 8
```

anijain2305 replied:
This seems like - we should have recompiled. But skipping all the guards made Dynamo think that it can reuse the previously compiled code. 

I guess, this flag is not so useful for you. What you want is to have more control over what to guard and what not to guard.

anijain2305 replied:
Converting this to WIP because I think we need more discussion on what kind of control is suitable for vLLM usecase.

anijain2305 committed:
Update on "[dynamo] Flag to skip all guards"


This is a requested feature by the users. What i have seen is users want something like `export` but not really `export`. They want the user friendliness of torch.compile but tight control over the fixed overheads.

The users are aware of the risks here and the responsibility/danger that comes with this flag. 


[ghstack-poisoned]

jansel reviewed (changes requested):
Add a test?
